email/model/jpa/src/main/java/com/blazebit/notify/email/model/jpa/AbstractEmailNotification.java

```diff
@@ -39,66 +36,73 @@
      * The value refers to a template name.
      */
-    public static final String SUBJECT_TEMPLATE_PARAMETER_NAME = "subjectTemplate";
+    public static final String SUBJECT_TEMPLATE_PARAMETER = "subjectTemplate";
+    /**
+     * The parameter name for the subject template processor key which is a plain {@link String}.
+     * The value refers to a template processor key.
+     */
+    public static final String SUBJECT_TEMPLATE_PROCESSOR_KEY_PARAMETER = "subjectTemplateProcessorKey";
     /**
      * The parameter name for the text body which is a plain {@link String}.
      * The value refers to the literal text body.
      */
-    public static final String BODY_TEXT_PARAMETER_NAME = "text";
+    public static final String BODY_TEXT_PARAMETER = "text";
     /**
      * The parameter name for the text body template which is a plain {@link String}.
      * The value refers to a template name.
      */
-    public static final String BODY_TEXT_TEMPLATE_PARAMETER_NAME = "textTemplate";
+    public static final String BODY_TEXT_TEMPLATE_PARAMETER = "textTemplate";
+    /**
+     * The parameter name for the html body template processor key which is a plain {@link String}.
+     * The value refers to a template processor key.
+     */
+    public static final String BODY_TEXT_TEMPLATE_PROCESSOR_KEY_PARAMETER = "textTemplateProcessorKey";
     /**
      * The parameter name for the html body which is a plain {@link String}.
      * The value refers to the literal html body.
      */
-    public static final String BODY_HTML_PARAMETER_NAME = "html";
+    public static final String BODY_HTML_PARAMETER = "html";
     /**
      * The parameter name for the html body template which is a plain {@link String}.
      * The value refers to a template name.
      */
-    public static final String BODY_HTML_TEMPLATE_PARAMETER_NAME = "htmlTemplate";
+    public static final String BODY_HTML_TEMPLATE_PARAMETER = "htmlTemplate";
+    /**
+     * The parameter name for the html body template processor key which is a plain {@link String}.
+     * The value refers to a template processor key.
+     */
+    public static final String BODY_HTML_TEMPLATE_PROCESSOR_KEY_PARAMETER = "htmlTemplateProcessorKey";
     /**
      * The parameter name for the attachments which is a collection of plain {@link String}s referring to
-     * template names. The templates are expected to be processed to a collection of {@link Attachment} by a
-     * {@link TemplateProcessor}.
+     * template processor keys. The template processors are expected to be processed to a collection of
+     * {@link Attachment} by a {@link TemplateProcessor}.
      */
-    public static final String ATTACHMENTS_PARAMETER_NAME = "attachments";
-
+    public static final String ATTACHMENT_TEMPLATE_PROCESSOR_KEYS_PARAMETER = "attachmentTemplateProcessorKeys";
     private static final long serialVersionUID = 1L;
 
-    private FromEmail from;
+    private F from;
     private Long fromId;
```

Comment:
I think you should remove the two fields and just leave a `public abstract AbstractFromEmail getFrom()` for a user to implement. AFAICS, you also don't need the type parameter. Users can covariantly override the contract.